src/openloco/windows/townwnd.cpp

```diff
@@ -1,31 +1,791 @@
+            {
+                case common::widx::caption:
+                    common::renameTownPrompt(self, widgetIndex);
+                    break;
+
+                case common::widx::close_button:
+                    WindowManager::close(self);
+                    break;
+
+                case common::widx::tab_town:
+                case common::widx::tab_population:
+                case common::widx::tab_company_ratings:
+                    common::switchTab(self, widgetIndex);
+                    break;
+            }
+        }
+
+        // 0x004996F6
+        static void on_resize(window* self)
+        {
+            // Call to sub_498E9B has been deliberately omitted.
+
+            self->set_size(gfx::ui_size_t(299, 172), gfx::ui_size_t(299, 327));
+
+            common::on_resize(self);
+        }
+
+        static void initEvents()
+        {
+            events.draw = draw;
+            events.on_mouse_up = on_mouse_up;
+            events.on_resize = on_resize;
+            events.on_update = common::update;
+            events.prepare_draw = prepare_draw;
+            events.text_input = common::text_input;
+        }
+    }
+
+    namespace company_ratings
+    {
+        static widget_t widgets[] = {
+            commonWidgets(340, 208, string_ids::title_town_local_authority),
+            widget_end(),
+        };
+
+        static window_event_list events;
+
+        // 0x00499761
+        static void prepare_draw(window* self)
+        {
+            common::prepare_draw(self);
+        }
+
+        // 0x004997F1
+        static void draw(window* self, gfx::drawpixelinfo_t* dpi)
+        {
+            self->draw(dpi);
+            common::drawTabs(self, dpi);
+
+            uint16_t xPos = self->x + 4;
+            uint16_t yPos = self->y + 46;
+            gfx::draw_string_494B3F(*dpi, xPos, yPos, colour::black, string_ids::local_authority_ratings_transport_companies);
+
+            xPos += 4;
+            yPos += 14;
+            auto town = townmgr::get(self->number);
+            for (company_id_t i = 0; i < std::size(town->company_ratings); i++)
+            {
+                if ((town->companies_with_rating & (1 << i)) == 0)
+                    continue;
+
+                int16_t rating = (std::clamp<int16_t>(town->company_ratings[i], -1000, 1000) + 1000) / 20;
+                string_id rank{};
+                if (rating >= 70)
+                    rank = string_ids::town_rating_excellent;
+                else if (rating >= 60)
+                    rank = string_ids::town_rating_good;
+                else if (rating >= 50)
+                    rank = string_ids::town_rating_average;
+                else if (rating >= 24)
+                    rank = string_ids::town_rating_poor;
+                else
+                    rank = string_ids::town_rating_appalling;
+
+                commonFormatArgs[0] = companymgr::get(i)->name;
+                commonFormatArgs[2] = rating;
```

Comment:
Any idea why this skips argument 1?

Reply:
No, I thought it strange, too, especially since `town_rating_company_percentage_rank` reads:

```
1839: "{COLOUR WINDOW_2}{STRINGID}: {COLOUR BLACK}{INT16}% ({STRINGID})"
```

So you'd expect it to not skip anything. But it does, as you've no doubt seen in IDA.